plugins/de.cau.cs.kieler.klighd.lsp/src/de/cau/cs/kieler/klighd/lsp/KGraphIncrementalDiagramGenerator.xtend

```diff
@@ -0,0 +1,753 @@
+        for( Pair<KEdge, List<SModelElement>> edgeAndParent : edgesToGenerate ) {
+            val edge = edgeAndParent.key
+            val parent = edgeAndParent.value
+            
+            // check if source and target of edge have already been generated
+            if (kGraphToSModelElementMap.containsKey(edge.target) && kGraphToSModelElementMap.containsKey(edge.source)){
+                val SEdge edgeElement = generateEdge(edge)
+                    if (edgeElement !== null) {
+                    parent.add(edgeElement)
+                    kGraphToSModelElementMap.put(edge, edgeElement)
+                    elementsToPostProcess.add(new Pair(edge, edgeElement))
+                    edgeElement.trace(edge)
+                }
+            } else {
+                remainingEdges.add(edgeAndParent)
+            }
+        }
+        // update edgesToGenerate
+        edgesToGenerate = remainingEdges
+    }
+
+    /**
+     * Translates all {@code ports} to SModelElements. Also handles tracing and mapping between
+     * KGraphElements and SModelElements.
+     */
+    private def List<SPort> createPorts(List<KPort> ports) {
+        val List<SPort> portElements = new ArrayList
+        for (port : ports) {
+            val SPort portElement = generatePort(port)
+            portElements.add(portElement)
+            kGraphToSModelElementMap.put(port, portElement)
+            elementsToPostProcess.add(new Pair(port, portElement))
+            portElement.trace(port)
+        }
+        return portElements
+    }
+
+    /**
+     * Translates all {@code labels} to SModelElements. Also handles tracing and mapping between
+     * KGraphElements and SModelElements.
+     */
+    private def List<SLabel> createLabels(List<KLabel> labels) {
+        val List<SLabel> labelElements = new ArrayList
+        for (label : labels) {
+            val SLabel labelElement = generateLabel(label, true)
+            labelElements.add(labelElement)
+            kGraphToSModelElementMap.put(label, labelElement)
+            elementsToPostProcess.add(new Pair(label, labelElement))
+            labelElement.trace(label)
+        }
+        return labelElements
+    }
+
+    /**
+     * Generates a trace for the {@code kElement}'s source EObject on the {@code sElement}. 
+     * The kElement must be synthesized by a KLighD synthesis before and must have its source EObject stored in the 
+     * {@link KlighdInternalProperties#MODEL_ELEMEMT} property.
+     */
+    private def void trace(SModelElement sElement, KGraphElement kElement) {
+        // The real model element that can be traced is the EObject that got synthesized in the
+        // {@link translateModel} function. That model element has to be stored in the properties during the 
+        // synthesis. Otherwise the tracing will not work
+        // FIXME: For large diagrams (expanded railway environment) the tracing alone
```

Comment:
This is the problem I was talking about earlier, see the fix on master to get rid of this FIXME.